### PR TITLE
core: Implement AffineSetAttr

### PR DESCRIPTION
**Pull request by PapyChacal**
Implement and tests AffineSetAttr parsing and printing.
Also enforce no imports from `xdsl.ir.affine.affine_X`, redirect to just `xdsl.ir.affine`

Thnaks @Groverkss for the heavy-lifting, because the rest is implemented, AffineSet is just a bunch of AffineExpr all over again.

Also remove @math-fehr's hack to allow parsing `affinet_set`s  as unregistered attributes!

### CODE REVIEW

File: tests/filecheck/mlir-conversion/with-mlir/affinet_set.mlir [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// RUN: xdsl-opt --allow-unregistered-dialect %s | mlir-opt --allow-unregistered-dialect --mlir-print-local-scope -mlir-print-op-generic | xdsl-opt --allow-unregistered-dialect
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/affinet_set.mlir:1]:
```suggestion
// RUN: xdsl-opt --allow-unregistered-dialect %s | mlir-opt --allow-unregistered-dialect --mlir-print-local-scope -mlir-print-op-generic | xdsl-opt --allow-unregistered-dialect | filecheck %s
```

File: xdsl/ir/affine/affine_set.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    """
+
+    num_dims: int
+    num_symbols: int
+    constraints: tuple[AffineConstraintExpr, ...]
+
+    def __str__(self) -> str:
+        # Create comma seperated list of dims.
+        dims = ["d" + str(i) for i in range(self.num_dims)]
+        dims = ", ".join(dims)
```
Comment (superlopuh) [inline xdsl/ir/affine/affine_set.py:41]: nit:
```suggestion
        dims = ", ".join("d" + str(i) for i in range(self.num_dims))
```

File: xdsl/parser/affine_parser.py [hunk 3]
```diff
@@ -121,6 +126,46 @@ def parse_expr() -> AffineExpr:
 
         return self.parse_comma_separated_list(self.Delimiter.PAREN, parse_expr)
 
+    # TODO: Extend to semi-affine maps
```
Comment (superlopuh) [inline xdsl/parser/affine_parser.py:129]: issue?

**Review by superlopuh — CHANGES_REQUESTED** [review]
very nice

File: xdsl/parser/affine_parser.py (hunk 3 above)
Comment (PapyChacal) [reply xdsl/parser/affine_parser.py:129]: I'll check if it doesn't exist, it dates from the affine_map work

Comment (math-fehr) [reply xdsl/parser/affine_parser.py:129]: And add the issue number on the TODO please ;)

File: xdsl/parser/attribute_parser.py [hunk 4]
```diff
@@ -51,6 +52,7 @@
 from xdsl.dialects.memref import MemRefType, UnrankedMemrefType
 from xdsl.ir import Attribute, Data, MLContext, ParametrizedAttribute
 from xdsl.ir.affine import AffineMap
+from xdsl.ir.affine.affine_set import AffineSet
```
Comment (math-fehr) [inline xdsl/parser/attribute_parser.py:55]:
```suggestion
from xdsl.ir.affine import AffineMap, AffineSet
```

File: xdsl/parser/attribute_parser.py [hunk 5]
```diff
@@ -688,47 +690,11 @@ def _parse_builtin_affine_map(self, _name: Span) -> AffineMapAttr:
...
-                    end_pos = token.span.end
-                    break
-                self._consume_token()
-
-        contents = self.lexer.input.slice(start_pos, end_pos)
-        assert contents is not None, "Fatal error in parser"
-
-        return attr_def(name.text, False, contents)
+    def _parse_builtin_affine_set(self, _name: Span) -> AffineSetAttr:
+        self.parse_characters("<", " in affine_map attribute")
```
Comment (math-fehr) [inline xdsl/parser/attribute_parser.py:694]:
```suggestion
        self.parse_characters("<", " in affine_set attribute")
```

File: xdsl/parser/attribute_parser.py [hunk 6]
```diff
@@ -688,47 +690,11 @@ def _parse_builtin_affine_map(self, _name: Span) -> AffineMapAttr:
...
+        affine_set = self.parse_affine_set()
+        self.parse_characters(">", " in affine_map attribute")
```
Comment (math-fehr) [inline xdsl/parser/attribute_parser.py:696]:
```suggestion
        self.parse_characters(">", " in affine_set attribute")
```

**Review by math-fehr — APPROVED** [review]

File: xdsl/dialects/builtin.py [hunk 7]
```diff
@@ -33,6 +33,7 @@
     TypeAttribute,
 )
 from xdsl.ir.affine import AffineMap
+from xdsl.ir.affine.affine_set import AffineSet
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/dialects/builtin.py:36]: Oh! Adding to the linter rules.

File: xdsl/parser/affine_parser.py (hunk 3 above)
Comment (PapyChacal) [reply xdsl/parser/affine_parser.py:129]: It exists; adding to the TODO
https://github.com/xdslproject/xdsl/issues/1087

File: xdsl/ir/affine/affine_set.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+@dataclass
+class AffineConstraintExpr:
+    kind: AffineConstraintKind
+    lhs: AffineExpr
+    rhs: AffineExpr
+
+    def __post_init__(self):
+        self.lhs = self.lhs - self.rhs
+        self.rhs = AffineExpr.constant(0)
```
Comment (superlopuh) [inline xdsl/ir/affine/affine_set.py:21]:
```suggestion
    def __init__(self, kind: AffineConstraintKind, lhs: AffineExpr, rhs: AffineExpr, *, canonicalize: bool = True):
        if canonicalize:
            lhs = self.lhs - self.rhs
            rhs = AffineExpr.constant(0)
        self.kind = kind
        self.lhs = lhs
        self.rhs = rhs
```

Can you please make it possible to opt out of this canonicalization?

**Review by superlopuh — APPROVED** [review]

**Pull request merged**